src/abstractdataframe/reshape.jl

```diff
@@ -399,3 +399,51 @@ function CategoricalArrays.CategoricalArray(v::RepeatedVector)
     res.refs = repeat(res.refs, inner = [v.inner], outer = [v.outer])
     res
 end
+
+
+Base.transpose(::AbstractDataFrame, args...; kwargs...) = MethodError("`transpose` not defined for `AbstractDataFrame`s. Try `permutedims` instead")
+
+"""
+    permutedims(df::AbstractDataFrame, src_namescol=1, dest_namescol=names(df)[src_namescol]; copycols::Bool=false, makeunique=false, promote_type::Bool=true)
+
+Transpose a `DataFrame`, such that rows become columns,
+and the column indexed by `src_namescol` becomes a header.
+
+By default, the type of resulting columns will depend on the promoted type of all input columns.
+Pass `promote_type=false` to make resulting column types depend on the elements of the originating rows,
+though note that this may be substantially slower.
```

Comment:
Actually I wonder whether we really need to add this argument now. Could you leave it out unless we have a reason to think the default isn't good enough? In general data frames should have concretely-types columns, and in that case promotion will give the best result.

Reply:
I think the reason for argument is, if columns being permuted have mixed types, you may have to chose between fast and concrete type:

```
julia> df = DataFrame(a=[randstring(5) for _ in 1:100], b=rand(100), c=repeat([:x, 1.], inner=50));

julia> pd_all = @btime permutedims(df, promote=:all);
  41.602 μs (410 allocations: 50.12 KiB)

julia> pd_none = @btime permutedims(df, promote=:none);
  89.565 μs (1070 allocations: 84.80 KiB)

julia> eltype.(eachcol(pd_none))
101-element Vector{DataType}:
 String
 Any
 Any
 Any
 Any
 Any
 ⋮
 Float64
 Float64
 Float64
 Float64
 Float64
 Float64

julia> eltype.(eachcol(pd_all))
101-element Vector{DataType}:
 String
 Any
 Any
 Any
 Any
 Any
 ⋮
 Any
 Any
 Any
 Any
 Any
 Any
```

Reply:
Yes but is that really common? If you permute your data frame, it sounds likely that the data is homogeneous.

Reply:
@kescobo - what @nalimilan means is that we can now implement only the default behaviour (leaving out the kwarg). The PR then can be merged, and later we can open another issue/PR to discuss how to best handle other options as this will be non-breaking.

Reply:
> Yes but is that really common?

@nalimilan For me, it's pretty common. There are some tools in my field that expect stuff in column-major format, and others that expect column-major. I often write my own version of this function to convert between the two.

But it's easy enough to concretize after the fact, so I'm fine with this.